osgar/drivers/tankmaria.py

```diff
@@ -0,0 +1,167 @@
+"""
+  Driver for a small tank Maria
+
+  Text Communication Protocol
+  Version 0 uses ASCII protocol between the host computer (JetsonNano) and Arduino motor controller.
+  There is no speed control implemented on Arduino at the moment, so raw PWM is used:
+     "<PWM left> <PWM right> <desired LED status>\n"
+  The value for PWM can in in -255..255.
+  LED status: 0=OFF, 1=green, 2=red.
+
+  The board regularly responds (20Hz?) with status line:
+      "<32bit unsigned int timestamp in milliseconds> <encoder left> <encoder right> <buttons mask>\n"
+  The encoders values are 32bit signed int and accumulative.
+  The buttons mask is currently not used (first remote tests are with Emergency STOP pressed).
+
+  The plan is to abandon this text protocol and use binary structures inspired by ATmega8board.
+"""
+
+import math
+
+from osgar.node import Node
+from osgar.bus import BusShutdownException
+
+
+ENC_SCALE = 0.00052
+WHEEL_DISTANCE = 0.4  # 50cm long, 40cm wide
+
+GREEN_LED = 1
+RED_LED = 2
+
+
+class MariaTimer(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register('tick')
+        self.sleep_time = config['sleep']
+
+    def run(self):
+        try:
+            while self.is_alive():
+                self.publish('tick', None)
+                self.sleep(self.sleep_time)
+
+        except BusShutdownException:
+            pass
+
+
+class RobotTankMaria(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        bus.register('pose2d', 'emergency_stop', 'encoders', 'raw')
+
+        # commands
+        self.desired_speed = 0.0  # m/s
+        self.desired_angular_speed = 0.0
+
+        # status
+        self.emergency_stop = None  # uknown state
+        self.pose = (0.0, 0.0, 0.0)  # x, y in meters, heading in radians (not corrected to 2PI)
+        self.buttons = None
+
+        self.verbose = False  # should be in Node
+        self.buf = b''
+        self.encoder = None
+        self.led = GREEN_LED
+        self.milisec = None
+
+    def send_pose(self):
+        x, y, heading = self.pose
+        self.publish('pose2d', [round(x*1000), round(y*1000),
+                                round(math.degrees(heading)*100)])
+
+
+    def compute_pose(self, left, right):
+        """Update internal pose with 'dt' step"""
```

Comment:
How does a time duration enter the calculation of the new pose?

Reply:
misleading copy & paste - fixed